File: frontend/src/constants/codeSpecialists.ts
import type { CodeSpecialist } from "../stores/workerStore";

export interface SpecialistInfo {
  id: CodeSpecialist;
  label: string;
  description: string;
  icon: string;
  color: string;
}

export const CODE_SPECIALISTS: SpecialistInfo[] = [
  {
    id: "general",
    label: "General",
    description: "Polyglot coder — works across any stack",
    icon: "M8 9l3 3-3 3m5 0h3M5 20h14a2 2 0 002-2V6a2 2 0 00-2-2H5a2 2 0 00-2 2v12a2 2 0 002 2z",
    color: "#10B981",
  },
  {
    id: "python_backend",
    label: "Python Backend",
    description: "FastAPI, Django, Flask — async Python, SQLAlchemy, pytest",
    icon: "M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-1 17.93c-3.95-.49-7-3.85-7-7.93 0-.62.08-1.21.21-1.79L9 15v1c0 1.1.9 2 2 2v1.93zm6.9-2.54c-.26-.81-1-1.39-1.9-1.39h-1v-3c0-.55-.45-1-1-1H8v-2h2c.55 0 1-.45 1-1V7h2c1.1 0 2-.9 2-2v-.41c2.93 1.19 5 4.06 5 7.41 0 2.08-.8 3.97-2.1 5.39z",
    color: "#3776AB",
  },
  {
    id: "fastify_backend",
    label: "Fastify Backend",
    description: "Fastify, Express — Node.js APIs, Prisma/Drizzle, TypeScript",
    icon: "M20 7l-8-4-8 4m16 0l-8 4m8-4v10l-8 4m0-10L4 7m8 4v10M4 7v10l8 4",
    color: "#000000",
  },
  {
    id: "nextjs",
    label: "Next.js",
    description: "App Router — RSC, server actions, middleware, ISR/SSG",
    icon: "M9 3v2m6-2v2M9 19v2m6-2v2M5 9H3m2 6H3m18-6h-2m2 6h-2M7 19h10a2 2 0 002-2V7a2 2 0 00-2-2H7a2 2 0 00-2 2v10a2 2 0 002 2zM9 9h6v6H9V9z",
    color: "#171717",
  },
  {
    id: "vite_react",
    label: "Vite + React",
    description: "Vite SPA — Zustand, TanStack, Tailwind, client-side routing",
    icon: "M13 10V3L4 14h7v7l9-11h-7z",
    color: "#646CFF",
  },
  {
    id: "dotnet_backend",
    label: ".NET Backend",
    description: "ASP.NET Core — minimal APIs, EF Core, DI, middleware",
    icon: "M5 12h14M5 12a2 2 0 01-2-2V6a2 2 0 012-2h14a2 2 0 012 2v4a2 2 0 01-2 2M5 12a2 2 0 00-2 2v4a2 2 0 002 2h14a2 2 0 002-2v-4a2 2 0 00-2-2m-2-4h.01M17 16h.01",
    color: "#512BD4",
  },
  {
    id: "dotnet_frontend",
    label: ".NET Frontend",
    description: "Blazor & Razor — component lifecycle, SignalR",
    icon: "M4 5a1 1 0 011-1h14a1 1 0 011 1v2a1 1 0 01-1 1H5a1 1 0 01-1-1V5zm0 8a1 1 0 011-1h6a1 1 0 011 1v6a1 1 0 01-1 1H5a1 1 0 01-1-1v-6zm12 0a1 1 0 011-1h2a1 1 0 011 1v6a1 1 0 01-1 1h-2a1 1 0 01-1-1v-6z",
    color: "#512BD4",
  },
];

export const SPECIALIST_MAP = Object.fromEntries(
  CODE_SPECIALISTS.map((s) => [s.id, s]),
) as Record<CodeSpecialist, SpecialistInfo>;

File: backend/axon/code_specialists.py
"""Code specialist definitions — expert coder profiles for worker agents.

Each specialist carries detection rules (to auto-identify from a codebase)
and points to a comprehensive instructions template that gets written into
the worker's vault as instructions.md.
"""

from __future__ import annotations

import logging
from enum import Enum
from pathlib import Path

logger = logging.getLogger(__name__)

# Instruction templates live alongside this module
INSTRUCTIONS_DIR = Path(__file__).parent / "specialist_instructions"


class CodeSpecialist(str, Enum):
    GENERAL = "general"
    PYTHON_BACKEND = "python_backend"
    FASTIFY_BACKEND = "fastify_backend"
    NEXTJS = "nextjs"
    VITE_REACT = "vite_react"
    DOTNET_BACKEND = "dotnet_backend"
    DOTNET_FRONTEND = "dotnet_frontend"


# ── Metadata ────────────────────────────────────────────────────────

SPECIALIST_LABELS: dict[CodeSpecialist, str] = {
    CodeSpecialist.GENERAL: "General",
    CodeSpecialist.PYTHON_BACKEND: "Python Backend",
    CodeSpecialist.FASTIFY_BACKEND: "Fastify Backend",
    CodeSpecialist.NEXTJS: "Next.js",
    CodeSpecialist.VITE_REACT: "Vite + React",
    CodeSpecialist.DOTNET_BACKEND: ".NET Backend",
    CodeSpecialist.DOTNET_FRONTEND: ".NET Frontend",
}

SPECIALIST_DESCRIPTIONS: dict[CodeSpecialist, str] = {
    CodeSpecialist.GENERAL: "General-purpose coder — works across any stack",
    CodeSpecialist.PYTHON_BACKEND: "FastAPI, Django, Flask — async Python, SQLAlchemy, Alembic, pytest",
    CodeSpecialist.FASTIFY_BACKEND: "Fastify, Express — Node.js APIs, Prisma/Drizzle, TypeScript",
    CodeSpecialist.NEXTJS: "Next.js App Router — RSC, server actions, middleware, ISR/SSG",
    CodeSpecialist.VITE_REACT: "Vite + React SPA — Zustand, TanStack, Tailwind, client-side routing",
    CodeSpecialist.DOTNET_BACKEND: "ASP.NET Core — minimal APIs, EF Core, DI, middleware pipeline",
    CodeSpecialist.DOTNET_FRONTEND: "Blazor & Razor — component lifecycle, SignalR, .NET frontend",
}

SPECIALIST_COLORS: dict[CodeSpecialist, str] = {
    CodeSpecialist.GENERAL: "#10B981",
    CodeSpecialist.PYTHON_BACKEND: "#3776AB",
    CodeSpecialist.FASTIFY_BACKEND: "#000000",
    CodeSpecialist.NEXTJS: "#171717",
    CodeSpecialist.VITE_REACT: "#646CFF",
    CodeSpecialist.DOTNET_BACKEND: "#512BD4",
    CodeSpecialist.DOTNET_FRONTEND: "#512BD4",
}

# ── Detection rules ─────────────────────────────────────────────────
# Each rule is a list of (file_or_glob, weight) pairs. The specialist
# with the highest total weight wins. Globs are checked with Path.glob
# against the codebase root.

_DETECTION_RULES: dict[CodeSpecialist, list[tuple[str, float]]] = {
    CodeSpecialist.NEXTJS: [
        ("next.config.js", 10),
        ("next.config.mjs", 10),
        ("next.config.ts", 10),
        ("app/layout.tsx", 5),
        ("app/layout.jsx", 5),
        ("app/page.tsx", 3),
        ("src/app/layout.tsx", 5),
        ("src/app/page.tsx", 3),
        ("pages/_app.tsx", 4),
        ("pages/_app.jsx", 4),
    ],
    CodeSpecialist.VITE_REACT: [
        ("vite.config.ts", 10),
        ("vite.config.js", 10),
        ("vite.config.mjs", 10),
        ("src/main.tsx", 3),
        ("src/main.jsx", 3),
        ("src/App.tsx", 2),
        ("index.html", 1),  # weak — many things have this
    ],
    CodeSpecialist.FASTIFY_BACKEND: [
        ("fastify", 0),  # package.json dependency — checked separately
        ("src/server.ts", 2),
        ("src/app.ts", 1),
        ("src/routes", 1),
        ("drizzle.config.ts", 3),
        ("prisma/schema.prisma", 3),
    ],
    CodeSpecialist.PYTHON_BACKEND: [
        ("pyproject.toml", 5),
        ("requirements.txt", 3),
        ("setup.py", 3),
        ("manage.py", 8),  # Django
        ("alembic.ini", 5),
        ("alembic/", 4),
        ("app/main.py", 3),  # FastAPI convention
        ("src/main.py", 2),
    ],
    CodeSpecialist.DOTNET_BACKEND: [
        ("*.csproj", 5),
        ("*.sln", 3),
        ("Program.cs", 4),
        ("appsettings.json", 3),
        ("Controllers/", 3),
        ("Startup.cs", 4),
    ],
    CodeSpecialist.DOTNET_FRONTEND: [
        ("_Imports.razor", 10),
        ("App.razor", 8),
        ("Pages/*.razor", 5),
        ("Components/*.razor", 5),
        ("wwwroot/", 2),
    ],
}


def detect_specialist(codebase_path: str) -> tuple[CodeSpecialist, dict[str, float]]:
    """Scan a codebase directory and return the best-matching specialist.

    Returns (specialist, scores) where scores maps each evaluated
    specialist to its total weight. If nothing matches, returns GENERAL.
    """
    root = Path(codebase_path)
    if not root.is_dir():
        return CodeSpecialist.GENERAL, {}

    scores: dict[str, float] = {}

    for specialist, rules in _DETECTION_RULES.items():
        total = 0.0
        for pattern, weight in rules:
            if weight == 0:
                continue  # placeholder for dependency checks
            target = root / pattern
            # Direct file/dir check
            if target.exists():
                total += weight
                continue
            # Glob check (for patterns with wildcards)
            if "*" in pattern:
                matches = list(root.glob(pattern))
                if matches:
                    total += weight
        scores[specialist.value] = total

    # Bonus: check package.json for framework-specific dependencies
    pkg_json = root / "package.json"
    if pkg_json.exists():
        try:
            import json
            pkg = json.loads(pkg_json.read_text(encoding="utf-8"))
            all_deps = {
                **pkg.get("dependencies", {}),
                **pkg.get("devDependencies", {}),
            }
            if "fastify" in all_deps:
                scores.setdefault(CodeSpecialist.FASTIFY_BACKEND.value, 0)
                scores[CodeSpecialist.FASTIFY_BACKEND.value] += 10
            if "express" in all_deps and "fastify" not in all_deps:
                scores.setdefault(CodeSpecialist.FASTIFY_BACKEND.value, 0)
                scores[CodeSpecialist.FASTIFY_BACKEND.value] += 6
            if "next" in all_deps:
                scores.setdefault(CodeSpecialist.NEXTJS.value, 0)
                scores[CodeSpecialist.NEXTJS.value] += 8
            if "vite" in all_deps:
                scores.setdefault(CodeSpecialist.VITE_REACT.value, 0)
                scores[CodeSpecialist.VITE_REACT.value] += 5
            if "react" in all_deps and "next" not in all_deps:
                scores.setdefault(CodeSpecialist.VITE_REACT.value, 0)
                scores[CodeSpecialist.VITE_REACT.value] += 2
        except Exception:
            pass

    # Bonus: check pyproject.toml for Python framework deps
    pyproject = root / "pyproject.toml"
    if pyproject.exists():
        try:
            content = pyproject.read_text(encoding="utf-8").lower()
            if "fastapi" in content:
                scores.setdefault(CodeSpecialist.PYTHON_BACKEND.value, 0)
                scores[CodeSpecialist.PYTHON_BACKEND.value] += 8
            if "django" in content:
                scores.setdefault(CodeSpecialist.PYTHON_BACKEND.value, 0)
                scores[CodeSpecialist.PYTHON_BACKEND.value] += 8
            if "flask" in content:
                scores.setdefault(CodeSpecialist.PYTHON_BACKEND.value, 0)
                scores[CodeSpecialist.PYTHON_BACKEND.value] += 6
            if "sqlalchemy" in content or "alembic" in content:
                scores.setdefault(CodeSpecialist.PYTHON_BACKEND.value, 0)
                scores[CodeSpecialist.PYTHON_BACKEND.value] += 3
        except Exception:
            pass

    # Bonus: check .csproj for Blazor/Razor indicators
    for csproj in root.glob("**/*.csproj"):
        try:
            content = csproj.read_text(encoding="utf-8").lower()
            if "microsoft.aspnetcore.components" in content or "blazor" in content:
                scores.setdefault(CodeSpecialist.DOTNET_FRONTEND.value, 0)
                scores[CodeSpecialist.DOTNET_FRONTEND.value] += 10
            elif "microsoft.aspnetcore" in content:
                scores.setdefault(CodeSpecialist.DOTNET_BACKEND.value, 0)
                scores[CodeSpecialist.DOTNET_BACKEND.value] += 5
        except Exception:
            pass
        break  # Only check first csproj

    # Pick winner
    if not scores or max(scores.values()) == 0:
        return CodeSpecialist.GENERAL, scores

    winner = max(scores, key=lambda k: scores[k])
    if scores[winner] < 3:  # minimum threshold
        return CodeSpecialist.GENERAL, scores

    return CodeSpecialist(winner), scores


def get_specialist_instructions(specialist: CodeSpecialist) -> str:
    """Load the instructions.md template for a specialist."""
    path = INSTRUCTIONS_DIR / f"{specialist.value}.md"
    if not path.exists():
        path = INSTRUCTIONS_DIR / "general.md"
    if not path.exists():
        return ""
    return path.read_text(encoding="utf-8")
